BedrockServer.cpp

```diff
@@ -2106,52 +2102,87 @@ void BedrockServer::_finishPeerCommand(unique_ptr<BedrockCommand>& command) {
+        for (auto portWrapper : portList) {
+            const unique_ptr<Port>& port = portWrapper.get();
 
-    // Try each port.
-    for (auto portWrapper : portList) {
-        const unique_ptr<Port>& port = portWrapper.get();
+            // Skip null ports (if the command or control port are closed).
+            if (!port) {
+                continue;
+            }
 
-        // Skip null ports (if the command or control port are closed).
-        if (!port) {
-            continue;
-        }
+            // Accept as many sockets as we can.
+            while (true) {
+                sockaddr_in addr;
+                int s = S_accept(port->s, addr, true); // Note that this sets the newly accepted socket to be blocking.
+
+                // If we got an error or no socket, done accepting for now.
+                if (s <= 0) {
+                    break;
+                }
 
-        // Accept as many sockets as we can.
-        while (true) {
-            sockaddr_in addr;
-            int s = S_accept(port->s, addr, true); // Note that this sets the newly accepted socket to be blocking.
+                // Otherwise create the object for this new socket.
+                SDEBUG("Accepting socket from '" << addr << "' on port '" << port->host << "'");
+                Socket socket(s, Socket::CONNECTED);
+                socket.addr = addr;
 
-            // If we got an error or no socket, done accepting for now.
-            if (s <= 0) {
-                break;
-            }
+                // If it came from a plugin, record that.
+                auto plugin = _portPluginMap.find(port);
+                if (plugin != _portPluginMap.end()) {
+                    socket.data = plugin->second;
+                }
 
-            // Otherwise create the object for this new socket.
-            SDEBUG("Accepting socket from '" << addr << "' on port '" << port->host << "'");
-            Socket socket(s, Socket::CONNECTED);
-            socket.addr = addr;
+                // And start up this socket's thread.
+                _outstandingSocketThreads++;
+                thread t;
+                bool threadStarted = false;
+                while (!threadStarted) {
+                    try {
+                        t = thread(&BedrockServer::handleSocket, this, move(socket), port == _controlPort, port == _commandPortPublic, port == _commandPortPrivate);
+                        threadStarted = true;
+                    } catch (const system_error& e) {
+                        // We don't care about this lock here from a performance perspective, it only happens when we
+                        // are unable to do any work anyway (i.e., we can't start threads).
+                        lock_guard<mutex> lock(_newSocketThreadBlockedMutex);
+                        if (!_newSocketThreadsBlocked) {
+                            // Block any new socket threads and warn.
+                            _newSocketThreadsBlocked = true;
```

Comment:
I know the comment below says we'll only ever hit this case with thousands of thread but if for some reason we hit it with fewer than 50 threads, we would be stuck, and if I'm not mistaken, we would have to restart the server. How about checking here for the count on `_outstandingSocketThreads` and if it's less than 50, log some other line instead of blocking.

Reply:
I'm trying to think of a good way to do this that doesn't just generate more race conditions.